### PR TITLE
feat: infra init

### PR DESCRIPTION
**Pull request by dixitaniket**
closes: #14

### CODE REVIEW

File: infra/go.mod [hunk 1]
```diff
@@ -0,0 +1,71 @@
+module contracts
```
Comment (adamewozniak) [inline infra/go.mod:1]:
```suggestion
module contract-relayer
```
?

File: infra/provision.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	nodeHostName := fmt.Sprintf("%s.%s.node.ojo.network", moniker, network.ChainID)
+	nodeHostNames = append(nodeHostNames, nodeHostName)
+
+	bootDisk := &compute.InstanceBootDiskArgs{
+		DeviceName: pulumi.String(fmt.Sprintf("%s-bootdisk", moniker)),
+		InitializeParams: &compute.InstanceBootDiskInitializeParamsArgs{
+			Image: pulumi.String("family/ubuntu-minimal-2204-lts"),
+			Type:  pulumi.String(network.NodeConfig.DiskType),
+			Size:  pulumi.Int(network.NodeConfig.DiskSizeGB),
+		},
```
Comment (adamewozniak) [inline infra/provision.go:51]: Are we spinning up a node here? I thought the goal should be to spin up a relayer to an existing contract quickly

File: infra/provision.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	}
+
+	conf := config.New(ctx, "")
+	sshPublic := conf.Require("sshpublic")
+	// TODO
+	sshPrivate := conf.RequireSecret("sshprivate").ApplyT(func(b64private string) (string, error) {
```
Comment (adamewozniak) [inline infra/provision.go:57]: This ssh key for any stack can be easily generated by https://github.com/umee-network/umee-infra/blob/main/lib/sshkeygen/ssh.go

File: infra/provision.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	relayerSpec := unit.UnitSpec{
+		Name:              techName,
+		Description:       fmt.Sprintf("%s daemon", techName),
+		User:              "blockchain",
+		BinaryInstallPath: fmt.Sprintf("/usr/local/bin/%s", network.LocalRelayerBinary),
+	}
+	relayerUnit := relayerSpec.ToUnit(fmt.Sprintf("/home/ubuntu/%s/relayer-config.toml", network.RelayerHomeFolderName))
+	// set environment for relayer keyring pass
+	environment := map[string]string{
+		"CW_RELAYER_PASS": "PASS",
```
Comment (adamewozniak) [inline infra/provision.go:158]: If we're creating a new wallet to use for each cw relayer we should also generate a password & put it as a secret in the pulumi config. We should consider these wallets funded & production

File: infra/provision.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+	}
+	relayerUnit.Environment = pulumi.ToStringMap(environment)
+
+	uploadCwRelayerBinary, err := remote.NewCopyFile(ctx, "relayer"+relayerUnit.Name+"-cp-cosmos-binary", &remote.CopyFileArgs{
+		Connection: conn,
+		// TODO: don't assume /usr/local/ as the base path (brittle); will work for now since we control action file, may not work on a particular devs machine
+		LocalPath:  pulumi.Sprintf("/usr/local/bin/%s", network.LocalRelayerBinary),
+		RemotePath: pulumi.Sprintf("/home/ubuntu/%s", network.LocalRelayerBinary),
```
Comment (adamewozniak) [inline infra/provision.go:166]: Let's set these as constants

File: infra/provision.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+	installCwRelayerBinary, err := remote.NewCommand(
+		ctx,
+		moniker+"-"+relayerUnit.Name+"wasm-install-cosmos-binary",
+		&remote.CommandArgs{
+			Connection: conn,
+			Create: pulumi.Sprintf(`
+						    set -e
+							sudo cp /home/ubuntu/%s /usr/local/bin/
+							sudo chmod a+x /usr/local/bin/%s
+						`, network.LocalRelayerBinary, network.LocalRelayerBinary),
```
Comment (adamewozniak) [inline infra/provision.go:181]: might want to move this over to scripts

File: infra/provision.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+	               if test -f "/tmp/STARTUP_FINISHED"; then
+	                 exit 0
+	               else
+	                 echo 'System startup script incomplete; sleeping 30 seconds...'
+	                 sleep 45
+	               fi
+	             done
+	
+	             echo 'Machine is not ready or system startup script did not complete (timeout)'
+	             exit 1
```
Comment (adamewozniak) [inline infra/provision.go:137]: might want to move this over to scripts file

**Review by adamewozniak — COMMENTED** [review]
this is looking good! just a few comments - we want to make sure each stack has a separate password for the wallet backend

**Comment by adamewozniak** [comment]
got eyes on this 👍

File: infra/provision.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		return string(privatebytes), nil
+	}).(pulumi.StringOutput)
+
+	instanceID := pulumi.IDInput(pulumi.ID(network.NodeConfig.InstanceID))
+	instance, err := compute.GetInstance(ctx, network.NodeConfig.InstanceName, instanceID, &compute.InstanceState{})
+
+	conn := remote.ConnectionArgs{
+		Host:       addrs.ToStringArrayOutput().Index(pulumi.Int(0)),
+		Port:       pulumi.Float64(22),
+		User:       pulumi.String("ubuntu"),
```
Comment (adamewozniak) [inline infra/provision.go:43]: consts please

File: infra/provision.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+		FolderMode:      pulumi.String("0755"),
+		FolderUser:      relayerUnit.User,
+		FolderGroup:     relayerUnit.User,
+		Triggers:        pulumi.Array{configPath, configBody},
+	}, pulumi.DependsOn([]pulumi.Resource{instance, installCwRelayerBinary}))
+	if err != nil {
+		return err
+	}
+
+	// start relayer demon, can also be removed as it already exists
```
Comment (adamewozniak) [inline infra/provision.go:201]:
```suggestion
	// start relayer daemon, can also be removed as it already exists
```

**Review by adamewozniak — APPROVED** [review]
a couple nitpicks but this looks good to me!
Let's test a release here before merging

**Comment by adamewozniak** [comment]
@dixitaniket please merge this in if your tests have gone through 👍

**Review by zarazan — APPROVED** [review]
Based on my current pulumi knowledge I think this looks great.

**Pull request merged**